wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift

```diff
@@ -117,124 +111,120 @@ public class MessageSender {
-        }) else {
-            return .failure(MessageSendError.failedToGenerateRequest)
-        }
 
-        let response = await httpClient.send(request)
+        let result = await apiProvider.messageAPI(apiVersion: apiVersion).sendProteusMessage(
+            payload: message,
+            conversationID: conversationID
+        )
 
-        return if response.result == .success {
+        return switch result {
+        case .success((let messageSendingStatus, let response)):
             managedObjectContext.performAndWait {
-                handleProteusSuccess(message: message, response: response, apiVersion: apiVersion)
+                handleProteusSuccess(
+                    message: message,
+                    messageSendingStatus: messageSendingStatus,
+                    response: response)
             }
-        } else {
-            await managedObjectContext.performAndWait({
-                handleProteusFailure(message: message, response: response, apiVersion: apiVersion)
-            }).flatMapAsync { retry in
-                if retry {
-                    await sessionEstablisher.establishSession(with: missingClients(), apiVersion: apiVersion)
-                        .mapError({ _ in
-                            MessageSendError.failedToGenerateRequest
-                        })
-                        .flatMapAsync { _ in
-                            await attemptToSendWithProteus(message: message, apiVersion: apiVersion)
+        case .failure(let networkError):
+            await managedObjectContext.performAndWait {
+                handleProteusFailure(message: message, networkError)
+            }.flatMapAsync { missingClients in
+                await sessionEstablisher.establishSession(with: missingClients, apiVersion: apiVersion)
+                    .mapError({ error in
+                        switch error {
+                        case .missingSelfClient: MessageSendError.missingPrecondition
+                        case .networkError(let networkError): MessageSendError.networkError(networkError)
                         }
-                } else {
-                    Swift.Result.failure(MessageSendError.gaveUpRetrying)
-                }
+                    })
+                    .flatMapAsync { _ in
+                        await attemptToSendWithProteus(message: message, apiVersion: apiVersion)
+                    }
             }
         }
     }
 
-    private func missingClients() -> Set<UserClient> {
+    private func missingClients() -> Set<QualifiedClientID> {
         return managedObjectContext.performAndWait {
-            ZMUser.selfUser(in: managedObjectContext).selfClient()?.missingClients ?? Set()
+            let clients = ZMUser.selfUser(in: managedObjectContext).selfClient()?.missingClients ?? Set()
+            return Set(clients.compactMap({ $0.qualifiedClientID })) // FIXME: [jacob] we can't do compact map
```

Comment:
is this still todo?

Reply:
No, it got solved in https://github.com/wireapp/wire-ios/pull/622